### PR TITLE
feat: per-step device memory tracking via ezpz.get_memory_metrics

### PR DESCRIPTION
**Pull request by saforem2**
## Summary

Adds per-step device memory tracking, condenses the History console
output into a single column-aligned summary line, and ships several
History/tplot fixes uncovered while testing the memory work.

### `ezpz.get_memory_metrics()` — new public API

```python
import ezpz

metrics = {"loss": loss.item(), "dtf": t}
metrics |= ezpz.get_memory_metrics(device, prefix="train/")
history.update(metrics)
```

Returns 4 keys (GiB) on CUDA / XPU, `{}` on CPU / MPS. Opt out via
`EZPZ_TRACK_MEMORY=0`. Wired into every example that calls
`History.update` (`test`, `vit`, `fsdp`, `fsdp_tp`, `diffusion`, `hf`,
`minimal`, `hf_trainer`).

Sibling helpers added to `ezpz.utils`:
`get_current_memory_allocated`, `get_current_memory_reserved`,
`reset_peak_memory_stats`. Existing
`get_max_memory_{allocated,reserved}` now return `0.0` on unsupported
devices instead of raising.

### Compact `key=value(±std)` console summary

`History.update()` now returns a single column-aligned line that
collapses the old 20+ token noise:

```
iter=180   loss=0.078(± 0.026) accuracy=0.984(±9.9e-3) lr=0.000059 dtf=0.014(±7.4e-4) memory=0.01/0.01GiB (0%)
```

Centralised in `ezpz.utils.format_compact_summary`:

- `/mean`, `/min`, `/max`, `/avg` companions dropped from the console
  (trackers still receive them)
- Std token right-aligned in a 6-char column so `(±0.070)` lines up
  under `(±5.1e-4)` and `(±  0.12)` across rows
- Counter-like keys (`iter`, `step`, `epoch`, `batch`, `idx`) bare and
  left-padded so the next field aligns
- Replicated hyperparameters (`lr`, `momentum`, `weight_decay`,
  `beta1/2`, `eps`, `clip_grad`, `warmup_steps`, …) bare — their std
  is always 0 across ranks
- Memory keys auto-collapsed to a single `memory=alloc/reserved (peak%)`
  token
- Std formatting: 2 sig figs, fixed-point ≥ 0.01, scientific
  notation < 0.01 with leading exponent zero stripped (`5.1e-4`,
  not `5.1e-04`)

### `ezpz.log.silence_noisy_loggers()` — new public helper

One-line setup to silence `urllib3`, `botocore`, `matplotlib.font_manager`,
etc. inside example scripts. Adopted by every example.

### History / tplot fixes (shaken out by the above)

- `History.log_metrics` condensed into a single line (was multi-line
  per-metric)
- Suppress noisy multi-pane tplot grids from stdout (`subplots` AND
  non-subplots branches)
- `tplot` clear: reset MAIN figure, not `_active` sub-pane
- `History`: unpack `overlay_order` as 3-tuples in metadata loop
- `format_memory_summary` review nits

## Test plan

- [x] `pytest tests/test_memory_metrics.py` — 56 tests pass (includes
      the 16 + 2 `TestFormatCompactSummary` cases covering counter
      padding, std alignment, known-constant exemption, std formatting,
      and the memory-key collapse)
- [x] `pytest tests/test_history.py tests/test_log.py tests/test_top_level_exports.py`
      — 50 more tests pass, no regressions
- [x] Single-rank `python -m ezpz.examples.test --model debug` smoke test
      verifies aligned output in the live training log
- [x] On real GPU (CUDA or XPU): `ezpz test` shows memory keys in
      summary log; W&B/MLflow shows 4 new series
- [x] `EZPZ_TRACK_MEMORY=0`: keys absent, no W&B series added

## Docs

- `docs/configuration.md`: `EZPZ_TRACK_MEMORY` added to the
  experiment-tracking env table
- `docs/history.md`: new "Console summary format" and "Device memory
  tracking" subsections; `EZPZ_TRACK_MEMORY` row in the History
  env-var table
- All new public APIs (`get_memory_metrics`, `format_compact_summary`,
  `silence_noisy_loggers`) have docstrings auto-rendered by mkdocstrings

## Release label

`release:minor` — new public API surface, no breaking changes. Existing
callers of `History.update()` continue to work; only the *shape* of the
returned summary string changes (denser, column-aligned).

### CODE REVIEW

File: src/ezpz/utils/__init__.py [hunk 1]
```diff
@@ -371,32 +375,127 @@ def normalize(name: str) -> str:
...
-    Args:
-        device (torch.device): The device to check memory allocation for.
+    Get the peak (max) memory allocated on the specified device since the
+    last reset. Bytes. Returns 0.0 on devices without memory tracking
+    (CPU, MPS).
     """
     if torch.cuda.is_available():
         return torch.cuda.max_memory_allocated(device)
-    elif torch.xpu.is_available():
+    if hasattr(torch, "xpu") and torch.xpu.is_available():
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/utils/__init__.py]: Fixed in 56689c3. The root cause was branching on `torch.cuda.is_available()` before checking device type — replaced with a small `_device_type()` helper that normalizes via `torch.device(...).type` and dispatches on that. All five memory helpers (`get_max_memory_allocated`, `get_max_memory_reserved`, `get_current_memory_allocated`, `get_current_memory_reserved`, `reset_peak_memory_stats`) now route correctly: CPU/MPS short-circuit to 0.0 / no-op before any backend call, even on CUDA-capable hosts.

Added a regression test (`test_cpu_helpers_not_routed_to_cuda_when_cuda_available`) that stubs every `torch.cuda.*` memory function to raise if called, then verifies each helper still returns 0.0 on `torch.device('cpu')`. That test would have caught the original bug.

[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/utils/__init__.py]: Same fix in 56689c3 — `get_max_memory_reserved` now dispatches via `_device_type()`. Covered by the same regression test.

File: src/ezpz/utils/__init__.py [hunk 2]
```diff
@@ -371,32 +375,127 @@ def normalize(name: str) -> str:
...
         try:
             return torch.xpu.max_memory_allocated(device)
-        except ImportError:
-            return -1.0
-    raise RuntimeError(f"Memory allocation not available for {device=}")
+        except (ImportError, AttributeError):
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/utils/__init__.py:424]: Same fix in 56689c3 — `reset_peak_memory_stats` now no-ops on CPU/MPS via the `_device_type()` dispatch, regardless of whether CUDA is available. Regression test covers this case too.

File: tests/test_memory_metrics.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    def test_env_var_unset_or_one_returns_metrics(self):
+        """Default (no env var) or '1' should NOT short-circuit."""
+        with patch.dict(os.environ, {"EZPZ_TRACK_MEMORY": "1"}):
+            # CPU still returns {} for a different reason (unsupported device);
+            # the point is that the env-var check didn't fire.
+            result = get_memory_metrics("cpu")
+            assert result == {}  # empty for device-reason, not env-reason
+
+
+class TestKeyShape:
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_memory_metrics.py:143]: Added in 56689c3 — new `TestXPUBackend` class with three tests:

1. `test_xpu_routes_to_xpu_apis`: monkeypatches `torch.cuda.is_available → False` and `torch.xpu.*` to fixed values, verifies `get_memory_metrics('xpu:0')` returns the expected dict.

2. `test_xpu_helpers_fallback_to_zero_on_attribute_error`: monkeypatches every `torch.xpu.*` memory function to raise `AttributeError` (simulating older torch builds with partial XPU support), verifies all four helpers return 0.0 without propagating.

3. `test_xpu_reset_peak_swallows_attribute_error`: same pattern for `reset_peak_memory_stats`.

All three skip gracefully on torch builds without `torch.xpu` (pre-2.4-ish). On macOS with current torch they run because `torch.xpu` is present as a stub module even on CPU-only builds.

File: tests/test_memory_metrics.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        """reset_peak_memory_stats() should not raise on CPU."""
+        reset_peak_memory_stats(torch.device("cpu"))  # no exception = pass
+
+    def test_get_memory_metrics_returns_empty_on_cpu(self):
+        """The top-level helper returns {} on CPU — no keys to mislead trackers."""
+        assert get_memory_metrics(torch.device("cpu")) == {}
+
+    def test_get_memory_metrics_returns_empty_on_mps(self):
+        """Same for MPS (Apple Silicon)."""
+        assert get_memory_metrics("mps") == {}
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_memory_metrics.py:45]: Added in 56689c3:

- `test_get_memory_metrics_accepts_torch_device_mps` — explicit `torch.device('mps')` instance.
- `test_get_memory_metrics_accepts_indexed_cpu_string` — `'cpu:0'` form.
- `test_default_device_resolution` (in `TestEnvVarOptOut`) — covers the `device=None` lazy resolution path by monkeypatching `ezpz.get_torch_device` to return CPU and asserting the result is `{}`.

The implementation is now also more robust to all these forms: instead of `str(device).startswith(('cpu','mps'))` it routes via `_device_type(device)` which normalizes through `torch.device(...).type`.

File: tests/test_memory_metrics.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        assert get_memory_metrics("cpu") == {}
+
+
+class TestEnvVarOptOut:
+    """EZPZ_TRACK_MEMORY=0 disables the feature even on supported devices."""
+
+    def test_env_var_zero_returns_empty(self):
+        with patch.dict(os.environ, {"EZPZ_TRACK_MEMORY": "0"}):
+            # Even with a device that would normally work, no keys.
+            assert get_memory_metrics("cuda:0") == {}
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_memory_metrics.py:105]: Strengthened in 56689c3. `TestEnvVarOptOut` now has four tests, all using mocked CUDA so the env-var behavior is what's being measured (not a confounding CPU short-circuit):

1. `test_env_var_zero_returns_empty_on_cuda` — `'0'` on a working CUDA device returns `{}`.
2. `test_env_var_unset_returns_metrics_on_cuda` — `patch.dict(os.environ, {}, clear=True)` on CUDA returns all 4 keys.
3. `test_env_var_one_returns_metrics_on_cuda` — `'1'` on CUDA returns all 4 keys.
4. `test_default_device_resolution` — covers the `device=None` path independently.

Previously the unset/'1' tests passed a CPU device, so they couldn't distinguish "env-var check fired" from "device-unsupported short-circuit fired" — exactly your concern. Fixed.

File: src/ezpz/utils/__init__.py [hunk 6]
```diff
@@ -371,32 +375,127 @@ def normalize(name: str) -> str:
     return name.strip("-")
 
 
-def get_max_memory_allocated(device: torch.device) -> float:
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/utils/__init__.py:374]: Addressed in 56689c3 — introduced `_device_type(device)` which centralizes the device-type detection. The five memory helpers now share a single dispatch pattern:

```python
def get_max_memory_allocated(device):
    dtype = _device_type(device)
    if dtype == 'cuda':
        return torch.cuda.max_memory_allocated(device)
    if dtype == 'xpu':
        try:
            return torch.xpu.max_memory_allocated(device)
        except (ImportError, AttributeError):
            return 0.0
    return 0.0
```

I went with this rather than your `_call_memory_backend(device, cuda_attr, xpu_attr, default)` form because the latter would require either `getattr(torch.cuda, cuda_attr)` lookups (string-based — loses type safety, harder to grep) or higher-order functions everywhere. The current shape is duplicative in the if-else structure but each function still reads as a direct, greppable call into `torch.cuda.foo` / `torch.xpu.foo`. Net 5 functions × ~6 lines each — and they were the bug-prone part, so I'd rather have them explicit than abstracted.

The `_device_type` extraction was the load-bearing change: it eliminates the duplicated "CUDA if globally available else XPU else 0" logic and replaces it with a single device-type lookup that all five helpers + `get_memory_metrics` use. The bug couldn't have happened with this in place.

File: src/ezpz/utils/__init__.py [hunk 7]
```diff
@@ -371,32 +375,127 @@ def normalize(name: str) -> str:
...
         return torch.cuda.max_memory_allocated(device)
-    elif torch.xpu.is_available():
+    if hasattr(torch, "xpu") and torch.xpu.is_available():
         try:
             return torch.xpu.max_memory_allocated(device)
-        except ImportError:
-            return -1.0
-    raise RuntimeError(f"Memory allocation not available for {device=}")
+        except (ImportError, AttributeError):
+            return 0.0
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/utils/__init__.py]: Fixed in 56689c3 via the `_device_type()` helper (see also the longer reply on Sourcery's complexity comment). All five memory helpers + `get_memory_metrics` now dispatch on `torch.device(device).type` instead of `torch.cuda.is_available()`. Regression test (`test_cpu_helpers_not_routed_to_cuda_when_cuda_available`) stubs every `torch.cuda.*` call to raise — proves CPU never routes there even when CUDA is "available".

File: src/ezpz/utils/__init__.py [hunk 8]
```diff
@@ -371,32 +375,127 @@ def normalize(name: str) -> str:
...
+        return {}
+    # Some callers pass torch.device('cuda', 0) which stringifies as
+    # 'cuda:0' — falls through. Others pass plain int; let the helpers
+    # handle device-coercion.
+
+    _GIB = 1024 ** 3
+    metrics = {
+        f"{prefix}mem_alloc":         get_current_memory_allocated(device) / _GIB,
+        f"{prefix}mem_peak_alloc":    get_max_memory_allocated(device) / _GIB,
+        f"{prefix}mem_reserved":      get_current_memory_reserved(device) / _GIB,
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/utils/__init__.py]: Addressed implicitly in 56689c3. The dict is now one key-value per line without alignment padding:

```python
metrics = {
    f'{prefix}mem_alloc': get_current_memory_allocated(device) / _GIB,
    f'{prefix}mem_peak_alloc': get_max_memory_allocated(device) / _GIB,
    f'{prefix}mem_reserved': get_current_memory_reserved(device) / _GIB,
    f'{prefix}mem_peak_reserved': get_max_memory_reserved(device) / _GIB,
}
```

No more manual column alignment for Ruff to reformat.

File: src/ezpz/utils/__init__.py [hunk 9]
```diff
@@ -371,32 +375,127 @@ def normalize(name: str) -> str:
...
-    Get the maximum memory allocated on the specified device.
-
-    Args:
-        device (torch.device): The device to check memory allocation for.
+    Get the peak (max) memory allocated on the specified device since the
+    last reset. Bytes. Returns 0.0 on devices without memory tracking
+    (CPU, MPS).
     """
     if torch.cuda.is_available():
         return torch.cuda.max_memory_allocated(device)
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/utils/__init__.py]: Fixed in 56689c3 (same as the Sourcery / Copilot reports of this same bug). Helpers now dispatch via a small `_device_type(device)` that normalizes through `torch.device(...).type` — so CPU/MPS inputs return 0.0 / `{}` regardless of what's globally available. The new `test_cpu_helpers_not_routed_to_cuda_when_cuda_available` regression test would have caught the original bug.
